core-api/src/main/java/com/optimizely/ab/config/audience/match/SemanticVersion.java

```diff
@@ -0,0 +1,348 @@
+                                    if (preRelease.length == 0 && targetVersion.preRelease.length > 0) {
+                                        result = 1; // No pre release wins over pre release
+                                    }
+                                    if (targetVersion.preRelease.length == 0 && preRelease.length > 0) {
+                                        result = -1; // No pre release wins over pre release
+                                    }
+                                    if (preRelease.length > 0 && targetVersion.preRelease.length > 0) {
+                                        int len = Math.min(preRelease.length, targetVersion.preRelease.length);
+                                        int count;
+                                        for (count = 0; count < len; count++) {
+                                            result = comparePreReleaseTag(count, targetVersion);
+                                            if (result != 0) {
+                                                break;
+                                            }
+                                        }
+                                        if (result == 0 && count == len) { // Longer version wins.
+                                            result = preRelease.length - targetVersion.preRelease.length;
+                                        }
+                                    }
+                                }
+                            } else {
+                                result = -1;
+                            }
+                        }
+                    }
+                } else {
+                    result = -1;
+                }
+            }
+        }
+        return result;
+    }
+
+    private int comparePreReleaseTag(int pos, SemanticVersion ov) {
+        Integer here = parseNumeric(preRelease[pos]);
+        Integer there = parseNumeric(ov.preRelease[pos]);
+
+        if (here != null && there == null) {
+            return -1; // Strings take precedence over numbers
+        }
+        if (here == null && there != null) {
+            return 1; // Strings take precedence over numbers
+        }
+        if (here == null) {
+            return (preRelease[pos].compareTo(ov.preRelease[pos])); // ASCII compare
+        }
+        return here.compareTo(there); // Number compare
+    }
+
+    // Parser implementation below
+
+    private final Integer[] vParts;
+    private final ArrayList<String> preParts;
+    private final ArrayList<String> metaParts;
+    private int errPos;
+    private final char[] input;
+
+    private boolean stateMajor() {
+        int pos = 0;
+        while (pos < input.length && input[pos] >= '0' && input[pos] <= '9') {
+            pos++; // match [0..9]+
+        }
+        if (pos == 0) { // Empty String -> Error
+            return false;
+        }
+        if (input[0] == '0' && pos > 1) { // Leading zero
```

Comment:
it means 0001.01.01 is wrong?

Reply:
yes a semantic version must not contain leading zeros.